### PR TITLE
Complete Breakup Of Calculation/DateTime Functions

### PR DESCRIPTION
**Pull request by oleibman**
In conjunction with parallel breakups happening in other areas of Calculation, this change breaks up all the DateTime functions into their own classes. All methods remaining in DateTime itself have a doc block deprecation notice, and consist only of stub code to call the replacement methods. Coverage of DateTime itself and all the replacement methods is 100%.

There is only one substantive change to the code (see next paragraph). Among the non-substantive changes, it now adopts the same parsing technique (throwing and catching exceptions) already in use in Engineering and MathTrig. Boolean parameters are allowed in lieu of numbers when Excel allows them. Most of the code changes involve refactoring due to the need to avoid Scrutinizer "complexity" failures in what it will consider to be new methods.

Issue #1936 was opened just as I was staging this. It is now fixed. One existing WORKDAY test was wrong (noted in a comment in the test data file), and a bunch of new tests are added.

I found it confusing to use DateTime as a node of the the class name since most of the methods invoke native DateTime methods. So, everything is moved to directory DateTimeExcel, and that is what is used in the class names.

There are several follow-up activities that I am planning to undertake if this PR is merged.

- ODS supports dates well before 1900. There are exactly 2 assertions for this functionality. More are needed (and some functions might have to change to accept this).
- WEEKDAY has some poorly documented extra options for "style" which are not yet implemented.
- Most tests have been changed to use a formula as entered on a spreadsheet rather than a direct call to the method which implements the formula. There are 3 exceptions at this time. WORKDAY and NETWORKDAYS, which include arrays as part of their parameters, are more complicated than most. YEARFRAC was just too large to deal with now.
- There are direct calls to the now-deprecated methods in both source code and tests, mostly in Financial code, but possibly in others as well. These need to be changed.
- Some constants, none "officially" documented, remain in the original class. These should be either deleted or marked deprecated. I wasn't sure if deprecation was even possible (or desirable), and did not want that to be something which would cause Scrutinizer to fail the change.

This is:

```
- [ ] a bugfix
- [ ] a new feature
- [x] code refactoring
```

Checklist:

- [x] Changes are covered by unit tests
- [x] Code style is respected
- [x] Commit message explains **why** the change is made (see https://github.com/erlang/otp/wiki/Writing-good-commit-messages)
- [ ] CHANGELOG.md contains a short summary of the change
- [ ] Documentation is updated as necessary

### Why this change is needed?
Long-term maintainability.

### CODE REVIEW

**Comment by MarkBaker** [comment]
> ODS supports dates well before 1900. There are exactly 2 assertions for this functionality. More are needed (and some functions might have to change to accept this).

I never understood why Microsoft set a 1900/1/1 start date for their calendar, or why they introduced the 1904 base date for the Mac calendar, or why they haven't subsequently allowed earlier dates

> WEEKDAY has some poorly documented extra options for "style" which are not yet implemented.
Most tests have been changed to use a formula as entered on a spreadsheet rather than a direct call to the method which implements the formula. There are 3 exceptions at this time. WORKDAY and NETWORKDAYS, which include arrays as part of their parameters, are more complicated than most. YEARFRAC was just too large to deal with now.

The original version of WEEKDAY() didn't have a style option at all (just the Sunday-Saturday 1-7), and when they did introduce it, it was only with options 1, 2 and 3

> There are direct calls to the now-deprecated methods in both source code and tests, mostly in Financial code, but possibly in others as well. These need to be changed.

I've spent a couple of days writing a small utility that generates a list of all the deprecated classes/methods/functions in  codebase, and then searches for calls that use them. It's nowhere near perfect, but did identify a couple of references to TRANSPOSE() that I'd left in LOOKUP(), and have already resolved; and
```
Call to deprecated method PhpOffice\PhpSpreadsheet\Calculation\MathTrig::numberOrNan
In file /PhpSpreadsheet/Calculation/MathTrig.php at line 1887
   Class: PhpOffice\PhpSpreadsheet\Calculation\MathTrig, method: builtinSQRT()
```
Knowing the way the DateTime and Financial functions are called, I suspect it will pick up a lot there once the refactoring is complete

> Some constants, none "officially" documented, remain in the original class. These should be either deleted or marked deprecated. I wasn't sure if deprecation was even possible (or desirable), and did not want that to be something which would cause Scrutinizer to fail the change.

I'd have to check on deprecation of constants; I'll take a look at that

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 1]
```diff
@@ -754,17 +754,17 @@ class Calculation
...
         'DATE' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'DATE'],
+            'functionCall' => [DateTimeExcel\Datefunc::class, 'funcDate'],
             'argumentCount' => '3',
         ],
         'DATEDIF' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'DATEDIF'],
+            'functionCall' => [DateTimeExcel\Datedif::class, 'funcDatedif'],
```
Comment (MarkBaker) [inline src/PhpSpreadsheet/Calculation/Calculation.php:762]: Perhaps use `DateDif` rather than `Datedif`: the capitalisation makes it easier to read at a glance.

File: src/PhpSpreadsheet/Calculation/DateTimeExcel/Datedif.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+     *
+     * @param mixed $startDate Excel date serial value, PHP date/time stamp, PHP DateTime object
+     *                                    or a standard date string
+     * @param mixed $endDate Excel date serial value, PHP date/time stamp, PHP DateTime object
+     *                                    or a standard date string
+     * @param string $unit
+     *
+     * @return int|string Interval between the dates
+     */
+    public static function funcDatedif($startDate, $endDate, $unit = 'D')
```
Comment (MarkBaker) [inline src/PhpSpreadsheet/Calculation/DateTimeExcel/Datedif.php:24]: Wondering if it's a good idea to create public class constants for the possible `$unit` values; that we can then use in the actual code (lines 49-54); and perhaps a private constant array of the values that can make input validation easier.

The same principle can be applied to fixed range of value arguments for other class/methods like WEEKDAY()

All part of improving the codebase as we go through this exercise

**Review by MarkBaker — COMMENTED** [review]
Great work, and it's really helping improve the codebase for the future, and catching some of those bugs with edge cases, or where there simply weren't tests before

I'm just glad that I never decided to have compatibility beyond the basic Excel/ODS/Gnumeric.... if I'd taken it to the level of different versions of Excel, it would have become impossible to support the changes that MS have made over the years

**Comment by MarkBaker** [comment]
Having just taken a quick look, it's possible to deprecate class constants, but not global constants. As almost all of our constants are public because the code was written before constant visibility existed in PHP, we should probably deprecate them all. I've moved the one general constant that exists in `Engineering` to a separate `Engineering\Constants` class, and deprecated the value in the `Engineering` class.
Constants that relate to a particular function, or constant arrays for lists of possible argument values should probably be deprecated in the old class, and moved as public (private for the arrays) to the appropriate function class.

I've already updated my DeprecationAnalysis script to include checking for class constants.

File: src/PhpSpreadsheet/Calculation/Calculation.php (hunk 1 above)
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Harder than it seems, because I'm developing on Windows with non-case-sensitive file names. Anticipating this difficulty, I tried several pushes till I found a scheme that works - rename to something entirely different, push, rename to desired name, push. So far, only DateDif is changed; I will work on the others later today.

Comment (MarkBaker) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: I'm on Windows myself (using Linux vagrant boxes for running the tests), so I've had that pain as well... and that's the only workround that I've ever found for the case-sensitive renames

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 3]
```diff
@@ -754,17 +754,17 @@ class Calculation
...
             'argumentCount' => '2,3',
         ],
         'DATEVALUE' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'DATEVALUE'],
+            'functionCall' => [DateTimeExcel\Datevalue::class, 'funcDatevalue'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 4]
```diff
@@ -919,7 +919,7 @@ class Calculation
         ],
         'EDATE' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'EDATE'],
+            'functionCall' => [DateTimeExcel\Edate::class, 'funcEdate'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 5]
```diff
@@ -934,7 +934,7 @@ class Calculation
         ],
         'EOMONTH' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'EOMONTH'],
+            'functionCall' => [DateTimeExcel\Eomonth::class, 'funcEomonth'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push. I preferred EoMonth to make it easier to translate from Excel name to PhpSpreadsheet name.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 6]
```diff
@@ -1760,7 +1760,7 @@ class Calculation
         ],
         'NETWORKDAYS' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'NETWORKDAYS'],
+            'functionCall' => [DateTimeExcel\Networkdays::class, 'funcNetworkdays'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push. I can't parse the name - is it supposed to be "Net Work Days" or "Network Days"? Reading the documentation doesn't give me any clue as to why either is correct. At any rate, went with your suggestion.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 7]
```diff
@@ -2569,12 +2569,12 @@ class Calculation
         ],
         'WEEKDAY' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'WEEKDAY'],
+            'functionCall' => [DateTimeExcel\Weekday::class, 'funcWeekday'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 8]
```diff
@@ -2569,12 +2569,12 @@ class Calculation
...
             'argumentCount' => '1,2',
         ],
         'WEEKNUM' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'WEEKNUM'],
+            'functionCall' => [DateTimeExcel\Weeknum::class, 'funcWeeknum'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push. Also IsoWeekNum.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 9]
```diff
@@ -2624,12 +2624,12 @@ class Calculation
...
         'YEAR' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'YEAR'],
+            'functionCall' => [DateTimeExcel\Year::class, 'funcYear'],
             'argumentCount' => '1',
         ],
         'YEARFRAC' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'YEARFRAC'],
+            'functionCall' => [DateTimeExcel\Yearfrac::class, 'funcYearfrac'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Resolved in latest push.

**Comment by oleibman** [comment]
Everything is resolved except for the constants. I'll work on them over the weekend.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 10]
```diff
@@ -2419,12 +2419,12 @@ class Calculation
...
         'TIME' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'TIME'],
+            'functionCall' => [DateTimeExcel\Time::class, 'funcTime'],
             'argumentCount' => '3',
         ],
         'TIMEVALUE' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'TIMEVALUE'],
+            'functionCall' => [DateTimeExcel\Timevalue::class, 'funcTimevalue'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Neglected to mark this resolved earlier.

File: src/PhpSpreadsheet/Calculation/Calculation.php [hunk 11]
```diff
@@ -2589,7 +2589,7 @@ class Calculation
         ],
         'WORKDAY' => [
             'category' => Category::CATEGORY_DATE_AND_TIME,
-            'functionCall' => [DateTime::class, 'WORKDAY'],
+            'functionCall' => [DateTimeExcel\Workday::class, 'funcWorkday'],
```
[the remark this answers is not in the document]
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/Calculation.php]: Neglected to mark this resolved earlier.

File: src/PhpSpreadsheet/Calculation/DateTimeExcel/Datedif.php (hunk 2 above)
Comment (oleibman) [reply src/PhpSpreadsheet/Calculation/DateTimeExcel/Datedif.php]: I followed your suggestion below of a separate Constants class. This seemed right because WeekDay will want to use the same constants as WeekNum when we get to implementing the new "style" values. All constants in DateTime are now marked deprecated following the model in Engineering.

**Comment by oleibman** [comment]
While awaiting your opinion on how to handle constants, I decided to work on the 3 test suites I didn't like - YEARFRAC, WORKDAY, and NETWORKDAYS. For YEARFRAC, I was able to figure out how to get it to use the existing data member while still letting the test result be calculated from a spreadsheet, not by a direct call. I wish I had thought of the technique I used here earlier - it would have simplified my changes to the other tests. Oh well.

And now that I had a more usable YEARFRAC test, I found another Excel 1900-calendar bug. When an empty cell is used as a date, some functions understandably treat it as 1900-01-01. And some, like YEARFRAC, treat it as the illegal date 1900-01-00! I had noticed a similar problem with the DAY function, but a single test was sufficient for that idiosyncrasy. YEARFRAC needs much more complicated logic to maintain bug-compatibility with this, and many more tests, all of which are now taken care of. I suspect I will have to review some other functions, at a later date.

I also figured out what I consider a good method for making a similarly non-disruptive change to the WORKDAY and NETWORKDAYS tests. There were only a small number of changes to the existing data, and the addition of only a small number of tests for each.

All constants in DateTime are now deprecated.

**Review by MarkBaker — APPROVED** [review]
A few small things, particularly in the use of those constants (I've a similar set in Financials, with some overlap)... but those can be resolved after both your changes and mine are merged to master
